### PR TITLE
Fix correct parsing of collections in AmazonLambdaRecorder

### LINKED ISSUE #40349: HandleRequestCollectionHelper for Quarkus Amazon Lambda
### Description

Hello.

We run several quarkus application as AWS Lambdas. Some are triggered by an AWS Api gateway and some by sqs or sns. A queue can collect some messages and send them via batching to a lambda. In JSON this would look like this if the message has a name and description attribute.

```
[{"name":"apple","description":"an apple a day..."},{"name":"strawberry","description":"red fruit"}]
```

A RequestHandler would look like this:

```
public class FruitsLambda implements RequestHandler<List<Fruit>, Void> {

    @Override
    public Void handleRequest(List<Fruit> fruit, Context context) {
        fruit.forEach(this::printName);
        return null;
    }

    private void printName(Fruit fruit) {
        Log.info(fruit.name());
        Log.info(fruit.description());
    }
}
```

When this runs sadly an exception occurs.

```
ERROR [io.qua.ama.lam.run.AbstractLambdaPollLoop] (Lambda Thread (TEST)) Failed to run lambda (TEST): java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class de.hamburml.quarkus.aws.lambda.Fruit (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; de.hamburml.quarkus.aws.lambda.Fruit is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @27b71f50)
```

The ObjectMapper cannot infer the correct type and as fallback uses a LinkedHashMap.

This can be solved with a little trick and this is my enhancement idea...

```
public abstract class HandleRequestCollectionHelper<INPUT_TYPE, OUTPUT_TYPE> implements RequestHandler<Object, OUTPUT_TYPE> {
    @Inject
    ObjectMapper objectMapper;

    @Override
    public OUTPUT_TYPE handleRequest(Object input, Context context) {
        INPUT_TYPE convertedInput = objectMapper.convertValue(input, this.getCollectionReference());
        this.handle(convertedInput, context);
        return null;
    }

    public abstract TypeReference<INPUT_TYPE> getCollectionReference();

    public abstract OUTPUT_TYPE handle(INPUT_TYPE input, Context context);

}
```

```
public class FruitsLambda2 extends HandleRequestCollectionHelper<List<Fruit>, Void> {

    @Override
    public TypeReference<List<Fruit>> getCollectionReference() {
        return new TypeReference<>() { };
    }

    @Override
    public Void handle(List<Fruit> input, Context context) {
        input.forEach(this::printName);
        return null;
    }

    private void printName(Fruit fruit) {
        Log.info(fruit.name());
        Log.info(fruit.description());
    }
}
```

With HandleRequestCollectionHelper the subclass is forced to implement getCollectionReference so that the ObjectMapper is able to convert the LinkedHashMap to the Collection-Type. If this is done inside a HandleRequestHelper this is hidden from the developer. Maybe getCollectionReference can also be saved with some reflection magic because the type should be already there in the INPUT_TYPE parameter.

HandleRequestCollectionHelper could be added to the amazon-lambda lib. Or maybe automatically used if a Collection-Type is used in first parameter of implementation of RequestHandler-Interface.

I have a simple example here https://github.com/hamburml/quarkus-lambda-handler-with-collection Under src/test are two packages, working and notworking.

Let me know what you think of that.

Thanks
Michael

### PR DESCRIPTION
**Pull request by hamburml**
test to make sure that the objectMapper parses the Input correct.

@patriot1burke thx for the hint.

I also adjusted the objectWriter but I am not sure if this is really necessary.

- Fixes: https://github.com/quarkusio/quarkus/issues/40349

### CODE REVIEW

**Review by geoand — APPROVED** [review]

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java [hunk 1]
```diff
@@ -53,12 +55,14 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
...
+        Type genericType = handlerMethod.getGenericParameterTypes()[0];
+        JavaType javaType = objectMapper.getTypeFactory().constructType(genericType);
         if (parameterType.equals(S3Event.class)) {
             objectReader = new S3EventInputReader(objectMapper);
         } else {
-            objectReader = new JacksonInputReader(objectMapper.readerFor(parameterType));
+            objectReader = new JacksonInputReader(objectMapper.readerFor(javaType));
         }
-        objectWriter = new JacksonOutputWriter(objectMapper.writerFor(handlerMethod.getReturnType()));
+        objectWriter = new JacksonOutputWriter(objectMapper.writerFor(javaType));
```
Comment (gsmet) [inline extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:65]: I'm surprised by this change? It used to use the return type and now AFAICS it uses the parameter type? Is it expected?

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks for the PR! I spotted something that looks a bit odd to me.

Also the tests are not passing.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java (hunk 1 above)
Comment (hamburml) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:65]: You are right. To be honest I simply adopted @patriot1burke comment in the issue and didnt gave it more thought. Just wanted to start the process of fixing this.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java [hunk 2]
```diff
@@ -53,12 +55,17 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
         ObjectMapper objectMapper = AmazonLambdaMapperRecorder.objectMapper;
         Method handlerMethod = discoverHandlerMethod(handlerClass);
         Class<?> parameterType = handlerMethod.getParameterTypes()[0];
+        Type genericParameterType = handlerMethod.getGenericParameterTypes()[0];
+        JavaType constructParameterType = objectMapper.getTypeFactory().constructType(genericParameterType);
+        Type genericReturnType = handlerMethod.getGenericReturnType();
+        JavaType constructReturnType = objectMapper.getTypeFactory().constructType(genericReturnType);
+
```
Comment (gsmet) [inline extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:62]: I think you'd rather move these lines just where they are needed. As typically in the `S3Event` case, some of this is useless.

File: extensions/amazon-lambda/deployment/src/test/java/io/quarkus/amazon/lambda/deployment/testing/LambdaHandlerET.java [hunk 3]
```diff
@@ -5,18 +5,14 @@
 
 import org.junit.jupiter.api.Test;
 
-import io.quarkus.test.junit.QuarkusTest;
-
-@QuarkusTest
```
Comment (gsmet) [inline extensions/amazon-lambda/deployment/src/test/java/io/quarkus/amazon/lambda/deployment/testing/LambdaHandlerET.java:10]: Oh, sorry, that I think we should keep.

**Review by gsmet — CHANGES_REQUESTED** [review]

**Review by gsmet — APPROVED** [review]
CI is happy so let's merge this.

Thanks for your contribution!

**Comment by gsmet** [comment]
@patriot1burke are you happy with it? If so, we can merge.

**Comment by gsmet** [comment]
@patriot1burke which reviews? I don't see any in this PR? Could you point me to where your comments are?

**Comment by geoand** [comment]
Maybe the review was not posted by accident (the GH UI tends to make this extremely easy to happen)?

**Comment by hamburml** [comment]
Maybe he means this comment in the linked Issue https://github.com/quarkusio/quarkus/issues/40349#issuecomment-2085262068 
To be honest I never had issues with the output of the lambda, so I did not do the same to the objectWriter.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java [hunk 4]
```diff
@@ -53,11 +55,15 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
...
         if (parameterType.equals(S3Event.class)) {
             objectReader = new S3EventInputReader(objectMapper);
+        } else if (Collection.class.isAssignableFrom(parameterType)) {
+            objectReader = new CollectionInputReader<>(objectMapper, handlerMethod);
         } else {
             objectReader = new JacksonInputReader(objectMapper.readerFor(parameterType));
         }
+
         objectWriter = new JacksonOutputWriter(objectMapper.writerFor(handlerMethod.getReturnType()));
```
[the remark this answers is not in the document]
Comment (patriot1burke) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:67]: Not seeing these?

**Review by patriot1burke — COMMENTED** [review]
See my reviews now?

**Comment by patriot1burke** [comment]
> Maybe he means this comment in the linked Issue [#40349 (comment)](https://github.com/quarkusio/quarkus/issues/40349#issuecomment-2085262068) To be honest I never had issues with the output of the lambda, so I did not do the same to the objectWriter.

That is one of the comments.  This PR is incomplete without supporting generic return type.

**Comment by hamburml** [comment]
@patriot1burke Now your comments are visible. Maybe tomorrow I will find some time. Not sure. Thanks.

**Comment by hamburml** [comment]
@patriot1burke I just pushed a new Test. Now I have a `RequestHandler<List<Person>, List<Person>>` and a `RequestHandler<List<String>, List<Person>>` lambda handler. They work out of the box, didn't need to change the writer. Any advice is appreciated, I do not know what I should do.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java [hunk 5]
```diff
@@ -53,11 +55,15 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
         ObjectMapper objectMapper = AmazonLambdaMapperRecorder.objectMapper;
         Method handlerMethod = discoverHandlerMethod(handlerClass);
         Class<?> parameterType = handlerMethod.getParameterTypes()[0];
+
         if (parameterType.equals(S3Event.class)) {
             objectReader = new S3EventInputReader(objectMapper);
+        } else if (Collection.class.isAssignableFrom(parameterType)) {
+            objectReader = new CollectionInputReader<>(objectMapper, handlerMethod);
```
[the remark this answers is not in the document]
Comment (hamburml) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:62]: IMHO, this looks consistent with the other code. I can change it if necessary.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamburml) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:67]: pls check new commit, i didnt need to to anything with the returnType and it worked for a collection.

**Comment by hamburml** [comment]
@patriot1burke I do not give up! I just renamed the tests and added new ones.
We now have a Test for an handleRequest which comes from an abstract class. 
We now have a Test where handleRequest gets as Input a Collection and as output a Collection.
The tests are green on my end. Please take a look.

**Comment by hamburml** [comment]
@patriot1burke anything pls?

**Comment by patriot1burke** [comment]
There's a conflict now with main.  Need to resolve conflicts

**Comment by hamburml** [comment]
> There's a conflict now with main. Need to resolve conflicts

I just rebased. I cant see any conflicts. Am I missing something here?

**Comment by hamburml** [comment]
This is really strange. Under GitHub actions I see the workflow run and it looks like it does not run. But I do not know why.

**Comment by patriot1burke** [comment]
I don't see the conflict anymore

**Review by patriot1burke — APPROVED** [review]

**Comment by gsmet** [comment]
And merged! Thanks a lot for your efforts!

**Comment by hamburml** [comment]
@gsmet could we backport this to current LTS?

**Comment by geoand** [comment]
+1 on backporting, but we'll also need https://github.com/quarkusio/quarkus/pull/41896 because the test that was added with this is flaky

**Pull request merged**